File: packages/loom/lsp/src/workspace.rs
//! Workspace-wide name index + per-document state.
//!
//! Tracks every open document and the cross-file index the LSP
//! request handlers query (spec §17): characters, traits, beats,
//! anchors, todos. Reparses on every edit and emits diagnostics.

use std::collections::HashMap;

use loom_parser::ast::{
    BodyItem, Conditional, Declaration, DeclarationKind, DialogueLine, Item, LoomFile,
};
use loom_parser::{parse, Diagnostic, Position, Severity, Span};
use lsp_types::{
    CompletionItem, DocumentSymbolResponse, GotoDefinitionResponse, Hover,
    PublishDiagnosticsParams, Range, Url,
};

use crate::{completion, definition, hover, symbols};

/// A single open document plus its latest parse.
#[derive(Debug)]
pub struct OpenDoc {
    pub text: String,
    pub file: LoomFile,
    pub diagnostics: Vec<Diagnostic>,
}

/// Cross-file index entry: one occurrence of a name.
#[derive(Clone, Debug)]
pub struct Occurrence {
    pub uri: Url,
    pub range: Range,
}

/// A captured `todo` fence body.
#[derive(Clone, Debug)]
pub struct Todo {
    pub uri: Url,
    pub range: Range,
    pub text: String,
}

/// Per-character snapshot we expose to hover.
#[derive(Clone, Debug)]
pub struct CharacterInfo {
    pub uri: Url,
    pub range: Range,
    /// Declared `is X, Y` mixins.
    pub mixins: Vec<String>,
    /// Raw indented body lines, one per source line.
    pub body: Vec<String>,
}

/// Per-beat snapshot used by hover + definition.
#[derive(Clone, Debug)]
pub struct BeatInfo {
    pub uri: Url,
    pub range: Range,
    pub name_range: Range,
    pub cast: Option<String>,
    pub setting: Option<String>,
}

/// In-memory project index.
#[derive(Debug, Default)]
pub struct Workspace {
    pub docs: HashMap<Url, OpenDoc>,
    pub characters: HashMap<String, CharacterInfo>,
    pub traits: HashMap<String, Occurrence>,
    pub beats: HashMap<String, Vec<BeatInfo>>,
    pub anchors: HashMap<String, Vec<Occurrence>>,
    pub todos: Vec<Todo>,
}

impl Workspace {
    pub fn new() -> Self {
        Self::default()
    }

    /// Insert or replace a document, reparse, rebuild the index.
    pub fn open(&mut self, uri: Url, text: String) {
        self.update(uri, text);
    }

    /// Replace a document's text and reparse.
    pub fn update(&mut self, uri: Url, text: String) {
        let (file, diagnostics) = parse(&text);
        self.docs.insert(
            uri,
            OpenDoc {
                text,
                file,
                diagnostics,
            },
        );
        self.rebuild_index();
    }

    /// Drop a document and rebuild.
    pub fn close(&mut self, uri: &Url) {
        self.docs.remove(uri);
        self.rebuild_index();
    }

    /// Latest diagnostics for a single document, packaged for
    /// `textDocument/publishDiagnostics`.
    pub fn diagnostics_for(&self, uri: &Url) -> Option<PublishDiagnosticsParams> {
        let doc = self.docs.get(uri)?;
        let diagnostics = doc.diagnostics.iter().map(to_lsp_diagnostic).collect();
        Some(PublishDiagnosticsParams {
            uri: uri.clone(),
            diagnostics,
            version: None,
        })
    }

    fn rebuild_index(&mut self) {
        self.characters.clear();
        self.traits.clear();
        self.beats.clear();
        self.anchors.clear();
        self.todos.clear();

        // Collect into temporary owned data so we don't borrow `self.docs`
        // mutably while iterating.
        let snapshot: Vec<(Url, LoomFile, String)> = self
            .docs
            .iter()
            .map(|(uri, doc)| (uri.clone(), doc.file.clone(), doc.text.clone()))
            .collect();

        for (uri, file, text) in snapshot {
            self.index_file(&uri, &file, &text);
        }
    }

    fn index_file(&mut self, uri: &Url, file: &LoomFile, text: &str) {
        for item in &file.items {
            match item {
                Item::Declaration(decl) => self.index_declaration(uri, decl),
                Item::Beat(beat) => {
                    let range = span_to_range(beat.span);
                    let name_range = name_range_in_text(text, beat.span, &beat.name);
                    let info = BeatInfo {
                        uri: uri.clone(),
                        range,
                        name_range,
                        cast: beat.contract.get("cast").map(|v| v.value.clone()),
                        setting: beat.contract.get("setting").map(|v| v.value.clone()),
                    };
                    self.beats.entry(beat.name.clone()).or_default().push(info);
                    for body in &beat.body {
                        self.index_body(uri, body);
                    }
                }
                Item::LetBinding(_) => {}
            }
        }
    }

    fn index_declaration(&mut self, uri: &Url, decl: &Declaration) {
        let range = span_to_range(decl.span);
        match decl.kind {
            DeclarationKind::Character | DeclarationKind::Role => {
                self.characters.insert(
                    decl.name.clone(),
                    CharacterInfo {
                        uri: uri.clone(),
                        range,
                        mixins: decl.mixin.clone(),
                        body: decl.body.iter().map(|l| l.text.clone()).collect(),
                    },
                );
            }
            DeclarationKind::Trait => {
                self.traits.insert(
                    decl.name.clone(),
                    Occurrence {
                        uri: uri.clone(),
                        range,
                    },
                );
            }
            _ => {}
        }
    }

    fn index_body(&mut self, uri: &Url, body: &BodyItem) {
        match body {
            BodyItem::Directive(d) => self.maybe_anchor(uri, &d.raw, d.span),
            BodyItem::DirectiveBlock(b) => {
                self.maybe_anchor(uri, &b.directive.raw, b.directive.span);
                for child in &b.body {
                    self.index_body(uri, child);
                }
            }
            BodyItem::Conditional(Conditional { arms, .. }) => {
                for arm in arms {
                    for child in &arm.body {
                        self.index_body(uri, child);
                    }
                }
            }
            BodyItem::Choice(c) => {
                for child in &c.body {
                    self.index_body(uri, child);
                }
            }
            BodyItem::Metadata(m) => {
                let trimmed = m.value.trim_start();
                if let Some(rest) = trimmed.strip_prefix("```todo") {
                    let text = rest
                        .trim_end_matches("```")
                        .trim_matches(|c: char| c == '\n' || c == '\r' || c == ' ')
                        .to_string();
                    self.todos.push(Todo {
                        uri: uri.clone(),
                        range: span_to_range(m.span),
                        text,
                    });
                }
            }
            BodyItem::Dialogue(block) => {
                for line in &block.lines {
                    if let DialogueLine::Directive(d) = line {
                        self.maybe_anchor(uri, &d.raw, d.span);
                    }
                }
            }
            _ => {}
        }
    }

    fn maybe_anchor(&mut self, uri: &Url, raw: &str, span: Span) {
        // `<anchor: name>` — store the name.
        let inner = raw.trim_start_matches('<').trim_end_matches('>').trim();
        if let Some(args) = inner.strip_prefix("anchor:") {
            let name = args.trim().to_string();
            if !name.is_empty() {
                self.anchors.entry(name).or_default().push(Occurrence {
                    uri: uri.clone(),
                    range: span_to_range(span),
                });
            }
        }
    }

    // ---------- request entry points ----------

    pub fn completion_at(&self, uri: &Url, pos: lsp_types::Position) -> Vec<CompletionItem> {
        completion::completion_at(self, uri, pos)
    }

    pub fn hover_at(&self, uri: &Url, pos: lsp_types::Position) -> Option<Hover> {
        hover::hover_at(self, uri, pos)
    }

    pub fn definition_at(
        &self,
        uri: &Url,
        pos: lsp_types::Position,
    ) -> Option<GotoDefinitionResponse> {
        definition::definition_at(self, uri, pos)
    }

    pub fn document_symbols(&self, uri: &Url) -> Option<DocumentSymbolResponse> {
        symbols::document_symbols(self, uri)
    }

    pub fn references_at(
        &self,
        uri: &Url,
        pos: lsp_types::Position,
    ) -> Vec<lsp_types::Location> {
        crate::references::references_at(self, uri, pos)
    }
}

/// Convert a parser [`Span`] to an LSP [`Range`]. Both use zero-based
/// line + UTF-8 byte column; UTF-16 conversion is deferred until we
/// hit a non-ASCII script. See `source.rs` for the rationale.
pub fn span_to_range(span: Span) -> Range {
    Range {
        start: position_to_lsp(span.start),
        end: position_to_lsp(span.end),
    }
}

pub fn position_to_lsp(p: Position) -> lsp_types::Position {
    lsp_types::Position {
        line: p.line,
        character: p.column,
    }
}

fn to_lsp_diagnostic(d: &Diagnostic) -> lsp_types::Diagnostic {
    lsp_types::Diagnostic {
        range: span_to_range(d.span),
        severity: Some(match d.severity {
            Severity::Error => lsp_types::DiagnosticSeverity::ERROR,
            Severity::Warning => lsp_types::DiagnosticSeverity::WARNING,
        }),
        code: Some(lsp_types::NumberOrString::String(d.code.id().to_string())),
        source: Some("loom".to_string()),
        message: d.message.clone(),
        ..Default::default()
    }
}

/// Locate a `name` substring inside a span's source text so we can
/// return a tight range for go-to-definition / symbol selection.
fn name_range_in_text(text: &str, span: Span, name: &str) -> Range {
    // Walk lines in the span and look for the first occurrence.
    let lines: Vec<&str> = text.lines().collect();
    for line_idx in span.start.line..=span.end.line {
        if let Some(line) = lines.get(line_idx as usize) {
            if let Some(col) = line.find(name) {
                let start = lsp_types::Position {
                    line: line_idx,
                    character: col as u32,
                };
                let end = lsp_types::Position {
                    line: line_idx,
                    character: (col + name.len()) as u32,
                };
                return Range { start, end };
            }
        }
    }
    span_to_range(span)
}

/// Extract the line of text at `pos.line` from `text`, or `""`.
pub fn line_at(text: &str, line: u32) -> &str {
    text.lines().nth(line as usize).unwrap_or("")
}

File: packages/loom/core/src/lsp/workspace.ts
//! Workspace-wide name index + per-document state.
//!
//! Tracks every open document and the cross-file index the LSP request
//! handlers query (spec §17): characters, traits, beats, anchors,
//! todos. Reparses on every edit and emits diagnostics.
//!
//! Mirrors the Rust `loom_lsp::Workspace`, with two TypeScript-side
//! shifts: documents are keyed by their raw URI string (no `url::Url`
//! normalization), and dialogue bodies are the unified `BodyItem[]`
//! list (the TS parser dropped the separate `DialogueLine` carrier), so
//! `indexBody` recurses into a dialogue's body like any other block.

import type {
  Beat,
  BodyItem,
  Declaration,
  LoomFile,
} from "../parser/ast.ts";
import { type Diagnostic as ParserDiagnostic, parse, parseMixinRef } from "../parser/index.ts";
import { Bundle, type LoomFileEntry } from "../runtime/bundle.ts";
import { compileModel } from "../runtime/sim/model.ts";
import type {
  CompletionItem,
  Diagnostic as LspDiagnostic,
  DocumentSymbolResponse,
  GotoDefinitionResponse,
  Hover,
  Location,
  Position,
  PublishDiagnosticsParams,
  Range,
} from "./types.ts";
import { DiagnosticSeverity } from "./types.ts";
import { stripPrefix, trimEndMatches, trimStartMatches } from "../parser/rust.ts";
import { findTokenSpans, lines, nameRangeInText, spanToRange, toLspDiagnostic } from "./util.ts";
import { completionAt } from "./completion.ts";
import { hoverAt } from "./hover.ts";
import { definitionAt } from "./definition.ts";
import { documentSymbols } from "./symbols.ts";
import { referencesAt } from "./references.ts";

/** A single open document plus its latest parse. */
export interface OpenDoc {
  text: string;
  file: LoomFile;
  diagnostics: ParserDiagnostic[];
}

/** Cross-file index entry: one occurrence of a name. */
export interface Occurrence {
  uri: string;
  range: Range;
}

/** A captured `todo` fence body. */
export interface Todo {
  uri: string;
  range: Range;
  text: string;
}

/** Per-character snapshot we expose to hover. */
export interface CharacterInfo {
  uri: string;
  /** The whole declaration span (used by hover). */
  range: Range;
  /** Tight range over just the declared name (used by go-to-definition). */
  nameRange: Range;
  /** Declared `is X, Y` mixins. */
  mixins: string[];
  /** Raw indented body lines, one per source line. */
  body: string[];
  /** Names of the class-owned `beat name(…)` blocks this character declares. */
  ownedBeats: string[];
}

/**
 * Per-trait index entry. Richer than a bare `Occurrence`: carries the
 * trait's parameter list (`TRAIT Scanner(beat)` → `["beat"]`) and the
 * names of the beats it ships (`beat greet(…)` blocks), so hover and
 * completion can surface a trait's shape without re-parsing.
 */
export interface TraitInfo {
  uri: string;
  range: Range;
  /** Tight range over just the trait name (used by go-to-definition). */
  nameRange: Range;
  params: string[];
  /** Shipped-beat names authored inside the trait body. */
  beats: string[];
}

/** Per-beat snapshot used by hover + definition. */
export interface BeatInfo {
  uri: string;
  range: Range;
  nameRange: Range;
  cast: string | null;
  setting: string | null;
}

/** In-memory project index. */
export class Workspace {
  docs = new Map<string, OpenDoc>();
  characters = new Map<string, CharacterInfo>();
  traits = new Map<string, TraitInfo>();
  beats = new Map<string, BeatInfo[]>();
  anchors = new Map<string, Occurrence[]>();
  todos: Todo[] = [];
  /** Cross-file project diagnostics keyed by owning document URI. */
  projectDiagnostics = new Map<string, LspDiagnostic[]>();

  /** Insert or replace a document, reparse, rebuild the index. */
  open(uri: string, text: string): void {
    this.update(uri, text);
  }

  /** Replace a document's text and reparse. */
  update(uri: string, text: string): void {
    const [file, diagnostics] = parse(text);
    this.docs.set(uri, { text, file, diagnostics });
    this.rebuildIndex();
  }

  /** Drop a document and rebuild. */
  close(uri: string): void {
    this.docs.delete(uri);
    this.rebuildIndex();
  }

  /** Drop every document (project switch / close) and rebuild empty. */
  reset(): void {
    if (this.docs.size === 0) return;
    this.docs.clear();
    this.rebuildIndex();
  }

  /**
   * Latest diagnostics for a single document, packaged for
   * `textDocument/publishDiagnostics`. Returns `null` if the document
   * is unknown.
   */
  diagnosticsFor(uri: string): PublishDiagnosticsParams | null {
    const doc = this.docs.get(uri);
    if (!doc) return null;
    const parser = doc.diagnostics.map(toLspDiagnostic);
    const project = this.projectDiagnostics.get(uri) ?? [];
    return { uri, diagnostics: [...parser, ...project] };
  }

  private rebuildIndex(): void {
    this.characters.clear();
    this.traits.clear();
    this.beats.clear();
    this.anchors.clear();
    this.todos = [];
    for (const [uri, doc] of this.docs) {
      this.indexFile(uri, doc.file, doc.text);
    }
    this.rebuildProjectDiagnostics();
  }

  /**
   * Batch replace many documents with a **single** index rebuild. Plain
   * `update` reparses + rebuilds the whole cross-file index (and recompiles
   * the throwaway diagnostics bundle) on every call, so pushing N files one
   * at a time is O(N²). `updateMany` parses each doc once, then rebuilds the
   * index a single time — used by the editor's whole-project indexer so
   * opening a project is one rebuild, not one-per-file.
   */
  updateMany(entries: Array<[string, string]>): void {
    if (entries.length === 0) return;
    for (const [uri, text] of entries) {
      const [file, diagnostics] = parse(text);
      this.docs.set(uri, { text, file, diagnostics });
    }
    this.rebuildIndex();
  }

  /**
   * Compile every open document into a throwaway [`Bundle`] and lift its
   * cross-file `projectDiagnostics` into per-URI LSP diagnostics. The
   * compile is wrapped in a try/catch so a malformed in-progress edit
   * never takes the index down — a compile failure just leaves the
   * project-diagnostic layer empty until the next keystroke fixes it.
   */
  private rebuildProjectDiagnostics(): void {
    this.projectDiagnostics.clear();
    if (this.docs.size === 0) return;
    const firstUri = this.docs.keys().next().value as string;
    const bundle = new Bundle();
    for (const [uri, doc] of this.docs) {
      const entry: LoomFileEntry = {
        path: uriToPath(uri),
        stem: stemOf(uri),
        qualifier: "",
        source: doc.text,
        file: doc.file,
        diagnostics: doc.diagnostics,
      };
      bundle.files.push(entry);
    }
    try {
      compileModel(bundle);
    } catch {
      // Defensive: a partial edit can trip a compile invariant. Never let
      // that break parser diagnostics / completion / hover.
      return;
    }
    for (const d of bundle.projectDiagnostics) {
      const [uri, diagnostic] = this.mapProjectDiagnostic(d, firstUri);
      const list = this.projectDiagnostics.get(uri);
      if (list) list.push(diagnostic);
      else this.projectDiagnostics.set(uri, [diagnostic]);
    }
  }

  /**
   * Map one cross-file [`ProjectDiagnostic`] to a `[uri, Diagnostic]`
   * pair. Character-owned diagnostics land on the declaring character's
   * document + range; project-wide ones (no character) attach to the
   * first document at the file head.
   */
  private mapProjectDiagnostic(
    d: Bundle["projectDiagnostics"][number],
    firstUri: string,
  ): [string, LspDiagnostic] {
    const head: Range = {
      start: { line: 0, character: 0 },
      end: { line: 0, character: 0 },
    };
    // Everything but the whole-project trio carries a `character`; resolve
    // it to the character's declaration site.
    const owned = "character" in d ? this.characters.get(d.character) : undefined;
    const uri = owned?.uri ?? firstUri;
    const range = owned?.range ?? head;
    let severity: DiagnosticSeverity = DiagnosticSeverity.Error;
    let message: string;
    switch (d.kind) {
      case "missingMainFile":
        message = "project has no `main.loom` entry file";
        break;
      case "entryBeatUnresolved":
        message = `entry beat \`${d.name}\` is not defined`;
        break;
      case "noEntryBeat":
        message = "project defines no entry beat";
        break;
      case "ambiguousSlot":
        message = `\`${d.character}\` inherits slot \`${d.prop}\` ambiguously from two traits`;
        break;
      case "requiredSlotUnfilled":
        message = `\`${d.character}\` leaves required slot \`${d.slot}\` unfilled`;
        break;
      case "requiredParamUnfilled":
        message = `trait \`${d.trait}\` requires an argument for \`${d.param}\``;
        break;
      case "unresolvedTraitArg":
        message = `\`${d.arg}\` (for \`${d.trait}.${d.param}\`) names no beat`;
        break;
      case "derivedBeatConflict":
        message = `two traits both ship a beat named \`${d.beat}\``;
        break;
      case "unfilledDerivedSlot":
        message = `derived beat \`${d.beat}\` has an unfilled slot \`${d.slot}\``;
        severity = DiagnosticSeverity.Warning;
        break;
    }
    return [uri, { range, severity, source: "loom", message }];
  }

  /**
   * The beat names an owner (CHARACTER / ROLE / TRAIT) makes reachable by
   * a qualified divert `-> owner.<beat>`: its own `beat` blocks plus the
   * beats shipped by every trait in its `is` chain (one level, via the
   * trait index). Used by completion inside `is Trait(…)` and after
   * `-> self.` / `-> Owner.`.
   */
  ownerBeatNames(name: string): string[] {
    const out = new Set<string>();
    const char = this.characters.get(name);
    if (char) {
      for (const b of char.ownedBeats) out.add(b);
      for (const mixin of char.mixins) {
        const trait = this.traits.get(parseMixinRef(mixin).name);
        if (trait) for (const b of trait.beats) out.add(b);
      }
    }
    const trait = this.traits.get(name);
    if (trait) for (const b of trait.beats) out.add(b);
    return [...out];
  }

  private indexFile(uri: string, file: LoomFile, text: string): void {
    for (const item of file.items) {
      switch (item.kind) {
        case "declaration":
          this.indexDeclaration(uri, item.value, text);
          break;
        case "beat":
          this.indexBeat(uri, item.value, text);
          break;
        case "letBinding":
          break;
      }
    }
  }

  private indexBeat(uri: string, beat: Beat, text: string): void {
    const range = spanToRange(beat.span);
    const nameRange = nameRangeInText(text, beat.span, beat.name);
    const info: BeatInfo = {
      uri,
      range,
      nameRange,
      cast: beat.contract.get("cast")?.value ?? null,
      setting: beat.contract.get("setting")?.value ?? null,
    };
    const entries = this.beats.get(beat.name);
    if (entries) entries.push(info);
    else this.beats.set(beat.name, [info]);
    for (const body of beat.body) {
      this.indexBody(uri, body);
    }
  }

  private indexDeclaration(uri: string, decl: Declaration, text: string): void {
    const range = spanToRange(decl.span);
    const nameRange = nameRangeInText(text, decl.span, decl.name);
    if (decl.kind === "character" || decl.kind === "role") {
      this.characters.set(decl.name, {
        uri,
        range,
        nameRange,
        mixins: [...decl.mixin],
        body: decl.body.map((l) => l.text),
        ownedBeats: decl.character?.beats.map((b) => b.name) ?? [],
      });
    } else if (decl.kind === "trait") {
      this.traits.set(decl.name, {
        uri,
        range,
        nameRange,
        params: decl.character ? [...decl.character.params] : [],
        beats: decl.character?.beats.map((b) => b.name) ?? [],
      });
    }
  }

  private indexBody(uri: string, body: BodyItem): void {
    switch (body.kind) {
      case "directive":
        this.maybeAnchor(uri, body.value.raw, spanToRange(body.value.span));
        break;
      case "directiveBlock":
        this.maybeAnchor(uri, body.value.directive.raw, spanToRange(body.value.directive.span));
        for (const child of body.value.body) this.indexBody(uri, child);
        break;
      case "conditional":
        for (const arm of body.value.arms) {
          for (const child of arm.body) this.indexBody(uri, child);
        }
        break;
      case "choice":
        for (const child of body.value.body) this.indexBody(uri, child);
        break;
      case "dialogue":
        // TS parser unified the dialogue body into `BodyItem[]` (no
        // separate `DialogueLine`), so anchors in dialogue directives —
        // and anything else — surface by recursing like any other block.
        for (const child of body.value.body) this.indexBody(uri, child);
        break;
      case "match":
        for (const arm of body.value.arms) {
          for (const child of arm.body) this.indexBody(uri, child);
        }
        break;
      case "eachVisit":
        for (const child of body.value.first) this.indexBody(uri, child);
        for (const child of body.value.then) this.indexBody(uri, child);
        for (const child of body.value.finally) this.indexBody(uri, child);
        break;
      case "afterMorph":
        for (const child of body.value.after) this.indexBody(uri, child);
        for (const child of body.value.otherwise) this.indexBody(uri, child);
        break;
      case "metadata": {
        // A metadata fence stores its *tail* (the text after the opening
        // ```), then any body lines — see the parser's `collectFence`.
        // A todo fence is therefore one whose tail word is `todo`. The
        // Rust LSP tested for a literal "```todo" prefix, which never
        // matched the backtick-free value the shared parser actually
        // produces (its `todos` index was dead); keying off the tail
        // populates it.
        const trimmed = body.value.value.replace(/^\s+/, "");
        const rest = stripPrefix(trimmed, "todo");
        if (rest !== null && (rest.length === 0 || " \t\r\n".includes(rest[0]!))) {
          const text = trimChars(rest, "\n\r ");
          this.todos.push({ uri, range: spanToRange(body.value.span), text });
        }
        break;
      }
      default:
        break;
    }
  }

  private maybeAnchor(uri: string, raw: string, range: Range): void {
    // `<anchor: name>` — store the name.
    const inner = trimEndMatches(trimStartMatches(raw, "<"), ">").trim();
    const args = stripPrefix(inner, "anchor:");
    if (args !== null) {
      const name = args.trim();
      if (name.length > 0) {
        const entry = this.anchors.get(name);
        if (entry) entry.push({ uri, range });
        else this.anchors.set(name, [{ uri, range }]);
      }
    }
  }

  // ---------- request entry points ----------

  completionAt(uri: string, pos: Position): CompletionItem[] {
    return completionAt(this, uri, pos);
  }

  hoverAt(uri: string, pos: Position): Hover | null {
    return hoverAt(this, uri, pos);
  }

  definitionAt(uri: string, pos: Position): GotoDefinitionResponse | null {
    return definitionAt(this, uri, pos);
  }

  documentSymbols(uri: string): DocumentSymbolResponse | null {
    return documentSymbols(this, uri);
  }

  referencesAt(uri: string, pos: Position): Location[] {
    return referencesAt(this, uri, pos);
  }

  /**
   * References to a *named* identifier without a position. Scans every
   * document for whole-word occurrences of `name`. Used by the editor's
   * References panel when a character / beat / world key is pinned in
   * the focus bus (no cursor — only the name).
   */
  referencesByName(name: string): Location[] {
    const out: Location[] = [];
    for (const [uri, doc] of this.docs) {
      lines(doc.text).forEach((lineText, lineNo) => {
        for (const [start, end] of findTokenSpans(lineText, name)) {
          out.push({
            uri,
            range: {
              start: { line: lineNo, character: start },
              end: { line: lineNo, character: end },
            },
          });
        }
      });
    }
    return out;
  }
}

/**
 * `s.trim_matches(|c| chars.contains(c))` — strip every leading and
 * trailing character that appears in the `chars` set.
 */
function trimChars(s: string, chars: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && chars.includes(s[start]!)) start += 1;
  while (end > start && chars.includes(s[end - 1]!)) end -= 1;
  return s.slice(start, end);
}

/** Strip a `scheme://` prefix off a document URI to get a path-ish string. */
function uriToPath(uri: string): string {
  const schemeEnd = uri.indexOf("://");
  return schemeEnd >= 0 ? uri.slice(schemeEnd + 3) : uri;
}

/** File stem for a URI — basename without the `.loom` suffix. */
function stemOf(uri: string): string {
  const path = uriToPath(uri);
  const base = path.split("/").pop() ?? path;
  return base.replace(/\.loom$/, "");
}
